### PR TITLE
[JUJU-154] [2.9] Refactor provisioning task to improve parallelism

### PR DESCRIPTION
**Pull request by achilleasa**
The previous implementation got triggered via a machine change event and
then attempted to parallelize the starting of the pending machines
within the batch. The same codepath also obtained a list of dead
machines and asked the provider to terminate a subset (depending on the
configured harvest mode) of them.

Both these operations are blocking so the provisioner task was unable to
process any other pending changes until it has finished processing the
current batch. In some substrates, machines can take up to a couple of
minutes to start. As a result, we could (reproducibly) bump into
pathological cases (LP1942100) where the machine creation was
effectively serialized and would significantly increase the time it took
to deploy complex bundles such as charmed k8s.

A similar problem is also seen when tearing down models (e.g.  during
our CI runs) where the provisioner has to wait for a set of machines to
be terminated by the environment before it has a chance to process the
next set of dead machines.

This PR changes the behavior of the provisioner task so as to
improve parallelism when starting/stopping instances. This is achieved
via the introduction of a worker pool into the provisioner task where
start/stop requests can be delegated thus freeing the provisioner task's
main loop to process any queued events.

This change comes with a few caveats which require additional logic in
the provisioner task:

1) As the starting/stopping of machines is handled by the worker pool,
provisioning-related errors are reported asynchronously. When any of the
pool workers detects an error, it queues it (in a buffered channel) and
then requests the pool to shutdown once all workers complete their work
and report any errors they may encounter. The main task loop can detect
when the pool is shutting down (by monitoring the channel returned via
the pool's Done() method) and invokes the pool's Close() method which
waits for the pool to shut down and returns any errors that got emitted
by the pool workers.

2) As the provisioner enumerates the dead machines every time it
processes a machine change (empty or not), it can observe the same dead
machines multiple times and try to stop the same machine twice. To avoid
this, the task keeps track of the machine IDs that are currently being
(asynchronously) terminated and ignores them if it encounters them again
as part of another processed change.

3) In a scenario where we run 'juju add-machine && juju kill-machine X',
the provisioner would normally attempt to start the machine and then try
to stop it while it is still being provisioned. This would result in an
error when the provisioner tries to set the status of the machine
(machine dead or not found) and cause the error to bubble up and cause
the provisioner task main loop to exit and get restarted.

To work around this problem, the provisioner tracks the machine IDs that
are currently being provisioned. If it detects that a machine which is
still provisioning is dead, it will not try to stop it immediately but
instead flag it so that we can defer stopping it until the machine comes
online. Once the StartInstance broker call returns, the code will check
if the machine was flagged for deferred stopping and immediately queue
a termination request for the machine.

Out of the box, each provisioner task (one **per model**) uses a pool of **16** workers.
However, this value can be customized on a per-model basis by changing the `num-provisioning-workers`
config option. The provisioner task gets notified when the config value changes and
resizes the pool (min workers is capped to 1) on demand.

## QA steps

You will need to test various scenarios on different substrates. For example:
1) A regular, bootstrap, deploy, (wait), remove scenario
2) Spin up multiple machines and kill some of them, e.g. `juju add-machine -n 10 && sleep 2 && juju remove-machine 1 3`.
3) Check that the AZ distribution logic has not been affected.
4) Test that `juju destroy-model` cleans up all machines
5) In HA mode, start some machines, kill some machines and then kill the leader; verify that we are not leaking instances.

To get better visibility in the provisioner internals: 

```
juju model-config logging-config="<root>=INFO;juju.worker.provisioner=DEBUG"
```

## Bug reference
https://bugs.launchpad.net/juju/+bug/1622813
https://bugs.launchpad.net/juju/+bug/1942100

### CODE REVIEW

File: environs/config/config.go [hunk 1]
```diff
@@ -75,6 +75,10 @@ const (
 	// ProvisionerHarvestModeKey stores the key for this setting.
 	ProvisionerHarvestModeKey = "provisioner-harvest-mode"
 
+	// NumProvisionWorkersKey is the key for the number of provision
```
Comment (wallyworld) [inline environs/config/config.go:78]: provisioner

**Review by wallyworld — COMMENTED** [review]
Quick initial comment - I need to look in more detail. I wonder if the pool size of 16 might be too large and could trigger rate limit errors on some clouds. I can't recall the exact batching used by the current provisioner task - I thought the main logic was just to fire off a go routine for each pending machine, but I can't remember how the pending machine batches were done. In any case, I wonder if we want to make the pool size smaller, or at least investigate the limits on the main clouds. Maybe we need a per cloud pool size.

**Comment by achilleasa** [comment]
@wallyworld IIRC the original code would just spin up a goroutine per machine ID in the incoming changes batch (no cap). Unless there is some magic to cap the changes list within the watcher I don't see how we could limit that. For example, try on 2.9/HEAD `juju add-machine -n 32`.

**Comment by hpidcock** [comment]
I tried to smash this on lxd with no problems

**Comment by hpidcock** [comment]
> I tried to smash this on lxd with no problems

seems to also work well with gap

**Comment by achilleasa** [comment]
!!build!!

File: worker/provisioner/provisioner_task_test.go [hunk 2]
```diff
@@ -621,14 +616,186 @@ func (s *ProvisionerTaskSuite) TestPopulateAZMachinesErrorWorkerStopped(c *gc.C)
...
+	// executed. Here is how everything is set up:
+	//
+	// We will register an event processed callback as a synchronization
+	// point.
+	//
+	// Machine 0 is alive but not yes started. Processing the first machine
+	// change will trigger a StartInstance call against the broker.  While
+	// in that call (i.e. the machine is still being started from the
+	// provisioner's perspective), we will set the machine as dead, queue a
+	// cahnge event for the same machine and wait until it has been
```
Comment (manadart) [inline worker/provisioner/provisioner_task_test.go:718]: "change"

File: worker/provisioner/provisioner_task.go [hunk 3]
```diff
@@ -328,13 +409,25 @@ func (task *provisionerTask) populateMachineMaps(ctx context.ProviderCallContext
...
+// that are currently being stopped or have been marked for deferred stopping
+// once they are online will be skipped.
 func (task *provisionerTask) pendingOrDead(
 	ids []string,
 ) (pending, dead []apiprovisioner.MachineProvisioner, err error) {
 	task.machinesMutex.RLock()
 	defer task.machinesMutex.RUnlock()
 	for _, id := range ids {
+		// Ignore machines that have been either queued for deferred
+		// stopping or they are currently sopping
```
Comment (manadart) [inline worker/provisioner/provisioner_task.go:422]: "stopping"

**Review by manadart — APPROVED** [review]
This looks good. I tested on AWS.
- Installed and tore down charmed k8s a couple of times.
- Created spaces and added batches of machines with different space constraints.
- Deployed ubuntu and added units in batches, ensuring AZ distribution.

**Pull request merged**